test/udruntime/include/updown.h

```diff
@@ -0,0 +1,1784 @@
+    if( global )
+      regions_to_use = &regions_global;
+    else
+      regions_to_use = &regions_local;
+    auto it = regions_to_use->find( ptr );
+    if( it == regions_to_use->end() || it->second.free ) {
+      UPDOWN_ERROR( "Trying to free pointer 0x%lX that is not in the regions"
+                    " or the region is free (double free?)",
+                    reinterpret_cast< uint64_t >( ptr ) );
+      return;
+    }
+    // merge left if free
+    if( it != regions_to_use->begin() && std::prev( it, 1 )->second.free ) {
+      uint64_t size = it->second.size;
+      it--;
+      it->second.size += size;
+      UPDOWN_INFOMSG( "Merging left 0x%lX to 0x%lX, adding %lu for a total of "
+                      "region with %lu",
+                      reinterpret_cast< uint64_t >( it->first ),
+                      reinterpret_cast< uint64_t >( std::next( it, 1 )->first ),
+                      size,
+                      it->second.size );
+      UPDOWN_INFOMSG( "Removing previous region at 0x%lX",
+                      reinterpret_cast< uint64_t >( it->first ) );
+      regions_to_use->erase( std::next( it, 1 ) );
+    }
+    // merge right if free
+    auto nextIt = std::next( it, 1 );
+    if( nextIt != regions_to_use->end() && nextIt->second.free ) {
+      uint64_t size = nextIt->second.size;
+      it->second.size += size;
+      UPDOWN_INFOMSG( "Merging right 0x%lX to 0x%lX, adding %lu for a total of "
+                      "region with %lu",
+                      reinterpret_cast< uint64_t >( it->first ),
+                      reinterpret_cast< uint64_t >( nextIt->first ),
+                      size,
+                      it->second.size );
+      UPDOWN_INFOMSG( "Removing previous region at 0x%lX",
+                      reinterpret_cast< uint64_t >( it->first ) );
+      regions_to_use->erase( nextIt );
+    }
+
+    // Check if there were no merges
+    if( it->first == ptr ) {
+#ifndef ASST
+      UPDOWN_INFOMSG( "No merges performed, just freeing 0x%lX = {%lu, %s}",
+                      reinterpret_cast< uint64_t >( it->first ),
+                      ( it->second.size ),
+                      ( it->second.free ) ? "Free" : "Used" );
+#else
+      if( it->second.free ) {
+        UPDOWN_INFOMSG( "No merges performed, just freeing 0x%lX = {%lu, Free}",
+                        reinterpret_cast< uint64_t >( it->first ),
+                        ( it->second.size ) );
+      } else {
+        UPDOWN_INFOMSG( "No merges performed, just freeing 0x%lX = {%lu, Used}",
+                        reinterpret_cast< uint64_t >( it->first ),
+                        ( it->second.size ) );
+      }
+#endif
+      it->second.free = true;
+    }
+  }
+};
+
+// moved from above to avoid inner classes
+template< class AllocationPolicy >
+class UDRuntime_t : public AllocationPolicy {
+private:
+  struct base_addr_t {
+    volatile ptr_t mmaddr;
+    volatile ptr_t spaddr;
+    volatile ptr_t ctrlAddr;
+    volatile ptr_t progAddr;
```

Comment:
Just FYI: `volatile` does not guarantee atomicity (`std::atomic` does), and `volatile`'s use is discouraged. I've even seen proposals to deprecate `volatile`.